src/Language/Spectacle/Interaction/CLI.hs

```diff
@@ -1,29 +1,124 @@
--- |
+{-# LANGUAGE OverloadedStrings #-}
+
+-- | This module exports the 'CLI' monad, an abstraction over command-line interactions such as emitting logs and
+-- messages from the model checker per options declared by a user.
+--
+-- @since 0.1.0.0
+module Language.Spectacle.Interaction.CLI
+  ( -- * The CLI Monad
+    CLI (CLI),
+    unCLI,
+
+    -- ** Running CLI
+    runCLI,
+
+    -- ** CLI Operations
+    cliPutDoc,
+
+    -- ** CLI Documents
+    cliResultDoc,
+
+    -- * CLI Context
+    ContextCLI (ContextCLI),
+    ctxOpts,
+    ctxHandle,
+
+    -- ** Construction
+    newContextCLI,
+  )
+where
+
+import Control.Monad (unless)
+import Control.Monad.IO.Class (MonadIO, liftIO)
+import Control.Monad.Reader (MonadReader, ReaderT, asks, runReaderT)
+import Prettyprinter (Doc, annotate, line, unAnnotate, (<+>))
+import Prettyprinter.Render.Terminal (AnsiStyle, Color (Green, Red), bold, color, hPutDoc)
+import System.IO (Handle, hClose)
+
+import Language.Spectacle.Interaction.Options (OptsCLI, isStdout, optsLogOutput, optsOnlyTrace)
+import qualified Language.Spectacle.Interaction.Options as Opts
+
+-- ---------------------------------------------------------------------------------------------------------------------
+
+-- | The 'CLI' monad is a @'ReaderT' 'IO'@ carrying context of command-line options.
+--
+-- @since 0.1.0.0
+newtype CLI a = CLI
+  {unCLI :: ReaderT ContextCLI IO a}
+  deriving stock (Functor)
+  deriving
+    (Applicative, Monad, MonadIO, MonadReader ContextCLI)
+    via ReaderT ContextCLI IO
+
+-- | Lower 'CLI' into 'IO' given command-line options.
 --
 -- @since 0.1.0.0
-module Language.Spectacle.Interaction.CLI where
+runCLI :: CLI a -> OptsCLI -> IO a
+runCLI cli opts = do
+  ctx <- newContextCLI opts
+  ret <- runReaderT (unCLI cli) ctx
+
+  unless (Opts.isStdout $ optsLogOutput opts) do
+    -- Close the handle to the log-output buffer created by
+    -- 'newContextCLI'/'handleFrom' after the command-line
+    -- interaction has been completed, if the handle was not
+    -- to System.IO.stdout
+    hClose (ctxHandle ctx)
```

Comment:
As mentioned during video-review, we can use the `managed` to take care of resource cleanup so we don't need to remember to do this, this feedback doesn't block this PR though and we will make a separate pass to clean that up.